crypto/kmac/kmac.c

```diff
@@ -0,0 +1,460 @@
+        OPENSSL_free(kctx);
+    }
+}
+
+static EVP_MAC_IMPL *kmac_new(void)
+{
+    EVP_MAC_IMPL *kctx;
+
+    if ((kctx = OPENSSL_zalloc(sizeof(*kctx))) == NULL
+        || (kctx->ctx = EVP_MD_CTX_new()) == NULL) {
+        kmac_free(kctx);
+        return NULL;
+    }
+    return kctx;
+}
+
+static int kmac_copy(EVP_MAC_IMPL *gdst, EVP_MAC_IMPL *gsrc)
+{
+    gdst->out_len = gsrc->out_len;
+    gdst->key_len = gsrc->key_len;
+    gdst->custom_len = gsrc->custom_len;
+    gdst->flags = gsrc->flags;
+    memcpy(gdst->key, gsrc->key, gsrc->key_len);
+    memcpy(gdst->custom, gsrc->custom, gdst->custom_len);
+
+    return EVP_MD_CTX_copy(gdst->ctx, gsrc->ctx);
+}
+
+/*
+ * The init() assumes that any ctrl methods are set beforehand for
+ * md, key and custom. Setting the fields afterwards will have no
+ * effect on the output mac.
+ */
+static int kmac_init(EVP_MAC_IMPL *kctx)
+{
+    EVP_MD_CTX *ctx = kctx->ctx;
+    unsigned char out[KMAC_MAX_BLOCKSIZE];
+    int out_len, block_len;
+
+    /* Only check for key here (setting the key is dependent on the digest) */
+    if (kctx->key_len == 0) {
+        EVPerr(EVP_F_KMAC_INIT, EVP_R_NO_KEY_SET);
+        return 0;
+    }
+    block_len = EVP_MD_CTX_block_size(ctx);
+
+    /* Set default custom string if it is not already set */
+    if (kctx->custom_len == 0)
+        (void)kmac_ctrl_str(kctx, "custom", "");
+
+    return bytepad(out, &out_len, kmac_string, sizeof(kmac_string),
+                   kctx->custom, kctx->custom_len, block_len)
+           && EVP_DigestUpdate(ctx, out, out_len)
+           && EVP_DigestUpdate(ctx, kctx->key, kctx->key_len);
+}
+
+static size_t kmac_size(EVP_MAC_IMPL *kctx)
+{
+    return kctx->out_len;
+}
+
+static int kmac_update(EVP_MAC_IMPL *kctx, const unsigned char *data,
+                       size_t datalen)
+{
+    return EVP_DigestUpdate(kctx->ctx, data, datalen);
+}
+
+static int kmac_final(EVP_MAC_IMPL *kctx, unsigned char *out)
+{
+    EVP_MD_CTX *ctx = kctx->ctx;
+    int lbits, len;
+    unsigned char encoded_outlen[KMAC_MAX_ENCODED_HEADER_LEN];
+
+    /* KMAC XOF mode sets the encoded length to 0 */
+    if ((kctx->flags & KMAC_FLAG_XOF_MODE) != 0)
+        lbits = 0;
+    else
+        lbits = kctx->out_len * 8;
+
+    return right_encode(encoded_outlen, &len, lbits)
+           && EVP_DigestUpdate(ctx, encoded_outlen, len)
+           && EVP_DigestFinalXOF(ctx, out, kctx->out_len);
+}
+
+/*
+ * The following Ctrl functions can be set any time before final():
+ *     - EVP_MAC_CTRL_SET_SIZE: The requested output length.
+ *     - EVP_MAC_CTRL_SET_XOF_MODE: If set, this indicates that
+ *           right_encoded(0) is part of the digested data, otherwise it uses
+ *           right_encoded(requested output length).
+
+ * All other Ctrl functions should be set before init().
+ * EVP_MAC_CTRL_SET_MD must be set before EVP_MAC_CTRL_SET_KEY.
+ */
+static int kmac_ctrl(EVP_MAC_IMPL *kctx, int cmd, va_list args)
```

Comment:
Was the (new) convention to call the varargs functions `_vctrl`?
I don't mind either way but we should remain consistent.

Reply:
I've called the public API that takes a `va_list` `EVP_MAC_vctrl`.  For the methods, it's assumed to take a `va_list` only, and I haven't really bothered in the ones I wrote since it's internal...  so uhmm [shrug]

Reply:
It doesnt have the v in other files either such as gmac and poly1305

Reply:
Cool, ignore me :)